### PR TITLE
[TargetLowering][AMDGPU][ARM][RISCV][X86] Teach SimplifyDemandedBits to combine (srl (sra X, C1), ShAmt) -> sra(X, C1+ShAmt)

### PR DESCRIPTION
**Pull request by topperc**
If the upper bits of the shr aren't demanded.

This helps with cases where the outer srl was originally an sra and was converted to a srl by SimplifyDemandedBits before it had a chance to combine with the inner sra. This can occur when the inner sra was part of a sign_extend_inreg expansion.

There are some regressions in ARM and Thumb2.

### CODE REVIEW

File: llvm/test/CodeGen/ARM/srem-seteq-illegal-types.ll [hunk 1]
```diff
@@ -244,8 +243,9 @@ define i1 @test_srem_pow2_setne(i6 %X) nounwind {
 ;
 ; ARM8-LABEL: test_srem_pow2_setne:
 ; ARM8:       @ %bb.0:
-; ARM8-NEXT:    sbfx r1, r0, #0, #6
-; ARM8-NEXT:    ubfx r1, r1, #9, #2
+; ARM8-NEXT:    lsl r1, r0, #26
```
Comment (topperc) [inline llvm/test/CodeGen/ARM/srem-seteq-illegal-types.ll:246]: Looks like ARM prefers to fold `asr #31` over folding an immediate. AArch64 on the other has a `let AddComplexity = 6` on the ANDWri pattern to priorite the immediate.

**Comment by topperc** [comment]
Adding a one use check fixed most of the regressions I was seeing

**Comment by dtcxzyw** [comment]
This patch breaks zext pattern:
```
; bin/llc -mtriple=riscv64 test.ll -o -
define i32 @func0000000000000000(i32 signext %0, i16 signext %1) nounwind {
entry:
  %2 = sdiv i16 %1, 2
  %3 = zext i16 %2 to i32
  %4 = add i32 %3, %0
  ret i32 %4
}
```
Before:
```
func0000000000000000:
        slli    a2, a1, 48
        srli    a2, a2, 63
        add     a1, a1, a2
        slli    a1, a1, 48
        srai    a1, a1, 48
        slli    a1, a1, 47
        srli    a1, a1, 48
        addw    a0, a1, a0
        ret
```
After:
```
func0000000000000000:
        slli    a2, a1, 48
        srli    a2, a2, 63
        add     a1, a1, a2
        slli    a1, a1, 48
        srai    a1, a1, 49
        lui     a2, 16
        addi    a2, a2, -1
        and     a1, a1, a2
        addw    a0, a1, a0
        ret
```
`and X, 2^16-1` can be lowered to `(X << 48) u>> 48`.

**Comment by topperc** [comment]
> This patch breaks zext pattern:
> 
> ```
> ; bin/llc -mtriple=riscv64 test.ll -o -
> define i32 @func0000000000000000(i32 signext %0, i16 signext %1) nounwind {
> entry:
>   %2 = sdiv i16 %1, 2
>   %3 = zext i16 %2 to i32
>   %4 = add i32 %3, %0
>   ret i32 %4
> }
> ```
> 
> Before:
> 
> ```
> func0000000000000000:
>         slli    a2, a1, 48
>         srli    a2, a2, 63
>         add     a1, a1, a2
>         slli    a1, a1, 48
>         srai    a1, a1, 48
>         slli    a1, a1, 47
>         srli    a1, a1, 48
>         addw    a0, a1, a0
>         ret
> ```
> 
> After:
> 
> ```
> func0000000000000000:
>         slli    a2, a1, 48
>         srli    a2, a2, 63
>         add     a1, a1, a2
>         slli    a1, a1, 48
>         srai    a1, a1, 49
>         lui     a2, 16
>         addi    a2, a2, -1
>         and     a1, a1, a2
>         addw    a0, a1, a0
>         ret
> ```
> 
> `and X, 2^16-1` can be lowered to `(X << 48) u>> 48`.

Thanks. I think I have an idea how to fix it. We can do this https://alive2.llvm.org/ce/z/kUowrW

**Comment by topperc** [comment]
What should we do about the ARM regressions? I verified that adding AddedComplexity to the patterns fixes the regression, but is disruptive to other tests.

File: llvm/test/CodeGen/ARM/srem-seteq-illegal-types.ll (hunk 1 above)
Comment (davemgreen) [reply llvm/test/CodeGen/ARM/srem-seteq-illegal-types.ll:246]: Considering these get optimized very differently by opt, and I don't see this coming up elsewhere I think it is OK. (I may look into it in the future to see if there is something we can do more generally to improve it).

Comment (topperc) [reply llvm/test/CodeGen/ARM/srem-seteq-illegal-types.ll:246]: Thanks @davemgreen

**Comment by topperc** [comment]
@RKSimon what do you think about this patch?

**Review by RKSimon — APPROVED** [review]
LGTM - as it looks like the AArch32/Thumb2 diffs are acceptable by the ARM team

**Comment by dtcxzyw** [comment]
Can you have a look at the following regression (extracted from velox):
```
; bin/llc -mtriple=riscv64 -mattr=+zba test.ll -o -

define ptr @func0000000000000004(ptr %0, i64 %1) nounwind {
entry:
  %2 = ashr i64 %1, 32
  %3 = lshr i64 %2, 6
  %4 = getelementptr i64, ptr %0, i64 %3
  ret ptr %4
}
```
Before:
```
func0000000000000004:
        srai    a1, a1, 32
        srli    a1, a1, 6
        sh3add  a0, a1, a0
        ret
```
After applying this patch:
```
func0000000000000004:
        srai    a1, a1, 35
        li      a2, -57
        srli    a2, a2, 3
        and     a1, a1, a2
        add     a0, a0, a1
        ret
```

**Comment by topperc** [comment]
> Can you have a look at the following regression (extracted from velox):
> 
> ```
> ; bin/llc -mtriple=riscv64 -mattr=+zba test.ll -o -
> 
> define ptr @func0000000000000004(ptr %0, i64 %1) nounwind {
> entry:
>   %2 = ashr i64 %1, 32
>   %3 = lshr i64 %2, 6
>   %4 = getelementptr i64, ptr %0, i64 %3
>   ret ptr %4
> }
> ```
> 
> Before:
> 
> ```
> func0000000000000004:
>         srai    a1, a1, 32
>         srli    a1, a1, 6
>         sh3add  a0, a1, a0
>         ret
> ```
> 
> After applying this patch:
> 
> ```
> func0000000000000004:
>         srai    a1, a1, 35
>         li      a2, -57
>         srli    a2, a2, 3
>         and     a1, a1, a2
>         add     a0, a0, a1
>         ret
> ```

Will do

**Pull request merged**